### PR TITLE
std/crypto/25519: add support for batch Ed25519 signature verification

### PR DESCRIPTION
**Pull request by jedisct1**
Batch verification allows verifying several signatures in a single operation, much faster than verifying signatures one-by-one. 

It uses random linear combinations to verify batch signatures, in a cofactored form to remain compatible with single signature verification.

### CODE REVIEW

File: lib/std/crypto/25519/ed25519.zig [hunk 1]
```diff
@@ -113,6 +113,75 @@ pub const Ed25519 = struct {
...
         } else |_| {}
     }
+
+    pub const BatchElement = struct {
+        sig: [signature_length]u8,
+        msg: []const u8,
+        public_key: [public_length]u8,
+    };
+
+    pub fn verify_batch(comptime count: usize, signature_batch: [count]BatchElement) !void {
```
Comment (daurnimator) [inline lib/std/crypto/25519/ed25519.zig:123]: should be `camelCase`?

File: lib/std/crypto/25519/scalar.zig [hunk 2]
```diff
@@ -150,11 +163,51 @@ pub inline fn clamp(s: *[32]u8) void {
...
+    return ScalarExpanded.fromBytes(a).mul(ScalarExpanded.fromBytes(b)).toBytes();
+}
+
 /// Return a*b+c (mod L)
 pub fn mulAdd(a: [32]u8, b: [32]u8, c: [32]u8) [32]u8 {
     return ScalarExpanded.fromBytes(a).mulAdd(ScalarExpanded.fromBytes(b), ScalarExpanded.fromBytes(c)).toBytes();
 }
 
+/// Return a*8 (mod L)
+pub fn mul8(s: [32]u8) [32]u8 {
```
Comment (daurnimator) [inline lib/std/crypto/25519/scalar.zig:177]: This seems a little... specific.
With only a single call site, why have this at all?
Alternatively, what if this was generic over some comptime known multiple?

File: lib/std/crypto/25519/scalar.zig [hunk 3]
```diff
@@ -150,11 +163,51 @@ pub inline fn clamp(s: *[32]u8) void {
...
+}
+
+/// Return a+b (mod L)
+pub fn add(a: [32]u8, b: [32]u8) [32]u8 {
+    return ScalarExpanded.fromBytes(a).add(ScalarExpanded.fromBytes(b)).toBytes();
+}
+
+/// Return -s (mod L)
+pub fn neg(s: [32]u8) [32]u8 {
+    comptime const fs: [64]u8 = field_size ++ [_]u8{0} ** 32;
```
Comment (daurnimator) [inline lib/std/crypto/25519/scalar.zig:192]: why `comptime` on this?

File: lib/std/crypto/25519/ed25519.zig (hunk 1 above)
Comment (jedisct1) [reply lib/std/crypto/25519/ed25519.zig:123]: Oh yes, it should be `camelCase`. Lemme change this right away.

File: lib/std/crypto/25519/scalar.zig (hunk 3 above)
Comment (jedisct1) [reply lib/std/crypto/25519/scalar.zig:192]: Not necessary, indeed.

File: lib/std/crypto/25519/scalar.zig (hunk 2 above)
Comment (jedisct1) [reply lib/std/crypto/25519/scalar.zig:177]: This is very specific. 8 is the cofactor of the groups these scalars represent elements of.

We only have one use case here, but multiplying a secret scalar per 8 is a common operation when using these curves. 

A similar operation is exposed as `clearCofactor()` in `edwards25519` and `curve25519`, but I didn't want to use that terminology for a finite field, where talking about cofactors wouldn't make sense.

**Review by daurnimator — APPROVED** [review]

**Comment by jedisct1** [comment]
With these changes, Zig becomes the only Ed25519 implementation that checks all the boxes from the recent [Taming the many EdDSAs](https://eprint.iacr.org/2020/1244.pdf) paper.

It achieves the strongest notion of security (until now, the only implementation achieving this was libsodium), but also does cofactored verification, with compatible batch verification (something that libsodium doesn't support yet).

**Comment by lithdew** [comment]
How would one identify the indices of all invalid signatures out of a batch?

**Comment by jedisct1** [comment]
> How would one identify the indices of all invalid signatures out of a batch?

You can't :) Well, you can, but that would be more expensive than verifying individual signatures.

Batch verification is typically used by blockchains that have to verify many signatures in blocks. Knowing whether a block is valid or not is all they need.

But even in other applications, if the vast majority of the signatures is assumed to be valid, you can gain speed by using batch verification, and then, only when the verification doesn't pass, perform individual verification on the same block,

**Comment by jedisct1** [comment]
Just a rebase after the WASI CI issue has been fixed.

**Comment by jedisct1** [comment]
Is LavaHut stuck?

**Comment by jedisct1** [comment]
Now using Barrett reduction and multiscalar multiplication for batch verification.

Benchmarks have been added as well. So, long story short:

Before:

```text
ed25519:      10469 verifications/s
```

After:

```text
ed25519:      35790 verifications/s (batch)
```

**Comment by jedisct1** [comment]
@andrewrk Do you think that could be merged? It also fixes an arithmetic overflow we previously had.

**Pull request merged**